code/game/machinery/vending/cm_vending.dm

```diff
@@ -1164,82 +1168,76 @@ GLOBAL_LIST_INIT(cm_vending_gear_corresponding_types_list, list(
-
-		if(vend_flags & VEND_UNIFORM_AUTOEQUIP)
-			// autoequip
-			if(istype(new_item, /obj/item) && new_item.flags_equip_slot != NO_FLAGS) //auto-equipping feature here
-				if(new_item.flags_equip_slot == SLOT_ACCESSORY)
-					if(user.w_uniform)
-						var/obj/item/clothing/clothing = user.w_uniform
-						if(clothing.can_attach_accessory(new_item))
-							clothing.attach_accessory(user, new_item)
-				else
-					user.equip_to_appropriate_slot(new_item)
-
-		if(vend_flags & VEND_TO_HAND)
-			if(user.client?.prefs && (user.client?.prefs?.toggle_prefs & TOGGLE_VEND_ITEM_TO_HAND))
-				if(vendor.Adjacent(user))
-					user.put_in_any_hand_if_possible(new_item, disable_warning = TRUE)
 	else
 		to_chat(user, SPAN_WARNING("ERROR: itemspec is missing. Please report this to admins."))
 		sleep(15)
 
 	vendor.stat &= ~IN_USE
 	vendor.update_icon()
 
+/proc/vendor_successful_vend_one(obj/structure/machinery/cm_vending/vendor, prod_type, mob/living/carbon/human/user, turf/target_turf, insignas_override)
+	var/obj/item/new_item
+	var/vend_flags = vendor.vend_flags
+	if(ispath(prod_type, /obj/item))
+		if(ispath(prod_type, /obj/item/weapon/gun))
+			new_item = new prod_type(target_turf, TRUE)
+		else
+			if(prod_type == /obj/item/device/radio/headset/almayer/marine)
+				prod_type = vendor.headset_type
+			else if(prod_type == /obj/item/clothing/gloves/marine)
+				prod_type = vendor.gloves_type
+			new_item = new prod_type(target_turf)
+		new_item.add_fingerprint(user)
+	else
+		new_item = new prod_type(target_turf)
+
+	if(vend_flags & VEND_UNIFORM_RANKS)
+		if(insignas_override)
+			var/obj/item/clothing/under/underclothes = new_item
+			//Gives ranks to the ranked
+			if(istype(underclothes) && user.wear_id && user.wear_id.paygrade)
+				var/rankpath = get_rank_pins(user.wear_id.paygrade)
+				if(rankpath)
+					var/obj/item/clothing/accessory/ranks/rank_insignia = new rankpath()
+					underclothes.attach_accessory(user, rank_insignia)
+
+	if(vend_flags & VEND_UNIFORM_AUTOEQUIP)
+		// autoequip
+		if(istype(new_item, /obj/item) && new_item.flags_equip_slot != NO_FLAGS) //auto-equipping feature here
+			if(new_item.flags_equip_slot == SLOT_ACCESSORY)
+				if(user.w_uniform)
+					var/obj/item/clothing/clothing = user.w_uniform
+					if(clothing.can_attach_accessory(new_item))
+						clothing.attach_accessory(user, new_item)
+			else
+				user.equip_to_appropriate_slot(new_item)
+
+	if(vend_flags & VEND_TO_HAND)
+		if(user.client?.prefs && (user.client?.prefs?.toggle_prefs & TOGGLE_VEND_ITEM_TO_HAND))
+			if(vendor.Adjacent(user))
+				user.put_in_any_hand_if_possible(new_item, disable_warning = TRUE)
+
 /proc/handle_vend(obj/structure/machinery/cm_vending/vendor, list/listed_products, mob/living/carbon/human/vending_human)
 	if(vendor.vend_flags & VEND_USE_VENDOR_FLAGS)
 		return TRUE
-	var/can_buy_flags = listed_products[4]
-	if(!(vending_human.marine_buy_flags & can_buy_flags))
+	var/buying_category = listed_products[4]
```

Comment:
This would probably involve changing a lot of files, but would be neat if instead of 4 it could be a define that says what index we're accessing containts e.g. listed_products[LISTED_PRODUCT_CATEGORY], etc. for the index usage.

Reply:
There's a problem with this in that the lists have different formats based on vendor type
I think this is out of scope for this PR in any case, would need to edit a lot more stuff and pray getting them all since there's no easy way to search for this